### ISSUE TITLE
Contour & Envoy memory usage

### ISSUE THREAD

**Issue by rosskukulinski** [heptio/contour#599]
I have been doing some performance testing to evaluate the memory usage of Contour & envoy as the number of IngressRoutes & Services increase.

I've noticed that Contour & Envoy memory increases considerably as the IngressRoutes & Services are being created, but a fresh-restart of Contour/Envoy results in a much-lower memory footprint.

Steps to reproduce:
*  I have a set of Pods (4x nginx) that match a label selector `app: nginx`
* Create IngressRoute object (ingress-0) that has a single Route to a Service `service-0`
* Create a service object (service-0) that uses a label selector `app:nginx`
* repeat for 0..2000
* Restart Contour & Envoy processes

My test ingressroute & service script repo is here: https://github.com/rosskukulinski/contour-envoy-memory-batch.

![screen shot 2018-07-31 at 1 49 21 pm](https://user-images.githubusercontent.com/2746479/43477400-193b32b8-94c9-11e8-824f-984fde3d48f5.png)

```

# services & # IngressRoutes | Envoy Memory (MiB)
-- | --
6.00 | 17
100.00 | 32
500.00 | 91
1000 | 326
1500 | 823
2000 | 1610
```

```
# services & # IngressRoutes | Contour Memory (MiB)
-- | --
6.00 | 8
100.00 | 37
500.00 | 375
1000 | 880
1500 | 1490
2000 | 2010
```

**Comment by stevesloka** [comment]
I'm curious @rosskukulinski if after you create the 2k objects, could you try deleting? Does the memory go down?

**Comment by rosskukulinski** [comment]
@stevesloka 
![screen shot 2018-07-31 at 3 13 46 pm](https://user-images.githubusercontent.com/2746479/43481597-62f4abd6-94d4-11e8-95f3-bf1570eb2059.png)

It took a while, but Contour did eventually start shedding memory

**Comment by rosskukulinski** [comment]
And I'm doing this with 1 contour, 5 envoys

**Comment by davecheney** [comment]
Hi Ross,

What’s the scale on this graph? I my testing I’ve not seen contour use more than 55mb of heap. 

Note, this is measured by contour using the go runtimes GODEBUG=gctrace=1 flag

**Comment by stevesloka** [comment]
@davecheney the left side is `GiB` and the bottom is time. I did the same test on my machine and Contour got to just over 7GiB memory. After some time it dropped down to around 3Gib which is where it's sitting now. 

I can try and re-run the test with the flags you mentioned. 

Here's a graph from my last run over 6 hours:
<img width="1551" alt="screen shot 2018-07-31 at 9 08 18 pm" src="https://user-images.githubusercontent.com/1048184/43495450-23c2c916-9506-11e8-9f3e-fac58f786a42.png">

**Comment by stevesloka** [comment]
Looks like (if I read the pprof correctly) is the memory usage is in the gRPC server:

```
(pprof) top
Showing nodes accounting for 862.22MB, 95.87% of 899.33MB total
Dropped 398 nodes (cum <= 4.50MB)
Showing top 10 nodes out of 37
      flat  flat%   sum%        cum   cum%
  645.88MB 71.82% 71.82%   829.25MB 92.21%  github.com/heptio/contour/vendor/google.golang.org/grpc/transport.(*http2Server).handleData
  183.36MB 20.39% 92.21%   183.36MB 20.39%  github.com/heptio/contour/vendor/google.golang.org/grpc/transport.(*recvBuffer).put
    9.12MB  1.01% 93.22%     9.18MB  1.02%  github.com/heptio/contour/internal/contour.(*cache).Values
    8.08MB   0.9% 94.12%     8.08MB   0.9%  github.com/heptio/contour/vendor/github.com/sirupsen/logrus.(*Entry).WithFields
    5.73MB  0.64% 94.76%    15.82MB  1.76%  github.com/heptio/contour/vendor/google.golang.org/grpc/transport.(*http2Server).operateHeaders
    4.60MB  0.51% 95.27%     4.60MB  0.51%  github.com/heptio/contour/vendor/google.golang.org/grpc/transport.(*decodeState).addMetadata (inline)
    2.21MB  0.25% 95.51%    27.01MB  3.00%  github.com/heptio/contour/internal/grpc.(*xdsHandler).stream
    1.78MB   0.2% 95.71%    28.94MB  3.22%  github.com/heptio/contour/vendor/google.golang.org/grpc.(*Server).processStreamingRPC
    1.31MB  0.15% 95.86%     5.07MB  0.56%  runtime.systemstack
    0.16MB 0.017% 95.87%    26.95MB  3.00%  github.com/heptio/contour/vendor/github.com/
```

**Comment by stevesloka** [comment]
I did the same test again looking at the go stats that get outputted from Contour, here are those graphs. The interesting piece is how the go routines top out at 10k. The graph shows a 3 hour window where I incrementally increased the Services/IngressRoutes, then the system sat idle to see how it reacts. 

<img width="1661" alt="screen shot 2018-08-02 at 4 50 05 pm" src="https://user-images.githubusercontent.com/1048184/43610445-26b6185c-9674-11e8-8d37-366fc7b5eb9a.png">

**Comment by davecheney** [comment]
My guess is grpc models each StreamEndpoints connection as a goroutine, if
each goroutine had a stack of 8kb and there are 10,000 of them, that's 2gb
of heap.

I'm not sure what can be done about this.

**Comment by rosskukulinski** [comment]
@rosskukulinski please reproduce to verify that envoy still has functionality after restarting contour.  In the original graph, it looks like both envoy & contour have low-memory usage after a contour restart.  That seems odd.

**Comment by rosskukulinski** [comment]
Also, I don't think this needs to hold back beta-3

**Comment by rosskukulinski** [comment]
I am not going to get to this before beta-3, moving to 0.6.0 milestone.

**Comment by rosskukulinski** [comment]
shifting to 0.8

**Comment by davecheney** [comment]
Nobody is available to work on this in 0.11. Punting to 0.12. Needs input from someone to verify if this is a problem.

**Comment by davecheney** [comment]
Duplicate of #499. This issue has useful information, but spanning two issues is not useful. I am closing this to concentrate on #499.
